### PR TITLE
Implement Connection Manager

### PR DESCRIPTION
**Pull request by Cohee1207**
It's a wrapper for STscript commands with a fancy interface. The idea is _not_ stolen TalOS-Reborn, all coincidences are coincidental.

> [!IMPORTANT] 
> IT IS NOT PRESET PRESETS!

image

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

### CODE REVIEW

**Review by RossAscends — COMMENTED** [review]
1. i would like to see this as core code, not an extension.
2. if tokenizer is being added into the profile attributes, the selector should be moved in here too (which would only be possible/sane if it's core code)

**Comment by Wolfsblvt** [comment]
1. Can be a core extension enabled by default. We have a few of those. Even QR, is not ***really*** core, if you think about it. Think this is a reasonable approach. Especially CC people likely won't care about this at all.
2. Advanced Formatting and Instruct is also not part of this tab, for the API. still part of the profile. Context too. All TC things actually. Why is the Tokenizer special to be moved?

**Comment by Cohee1207** [comment]
1. I think at this point adding new core features is always a tough decision, because of xkcd 1172 (see below). So I'd want to give an ability to disable this for those who don't need it.
2. I don't think the API pane is a proper place for the tokenizer since you can switch between Together/Mancer/OpenRouter/local and would still have the same Llama 3 model running under the hood with the same tokenizer and formatting. In the ideal world, API endpoints would provide the methods of tokenization and we would not have to worry about tokens anymore. By we're not in the ideal world.

xkcd 1172

File: public/scripts/slash-commands.js [hunk 1]
```diff
@@ -2299,7 +2299,7 @@ async function echoCallback(args, value) {
     if (args.extendedTimeout && !isNaN(parseInt(args.extendedTimeout))) options.extendedTimeOut = parseInt(args.extendedTimeout);
     if (isTrueBoolean(args.preventDuplicates)) options.preventDuplicates = true;
     if (args.cssClass) options.toastClass = args.cssClass;
-    if (args.escapeHtml !== undefined) options.escapeHtml = isTrueBoolean(args.escapeHtml);
+    options.escapeHtml = args.escapeHtml !== undefined ? isTrueBoolean(args.escapeHtml) : true;
```
Comment (Wolfsblvt) [inline public/scripts/slash-commands.js]: This is not needed, we have set `escapeHtml` set to default globally. But I don't object, it doesn't change anything.

File: public/scripts/extensions/connection-manager/index.js [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        callback: async () => {
+            const selectedProfile = extension_settings.connectionManager.selectedProfile;
+            const profile = extension_settings.connectionManager.profiles.find(p => p.id === selectedProfile);
+            if (!profile) {
+                return '';
+            }
+            await updateConnectionProfile(profile);
+            await renderDetailsContent(details, detailsContent);
+            saveSettingsDebounced();
+        },
```
Comment (Wolfsblvt) [inline public/scripts/extensions/connection-manager/index.js:563]: You still need to return empty string on callbacks. Good practice, and such.
Or return something meaningful, like the profile name. We do that on a lot of slash commands even on set.

File: public/scripts/extensions/connection-manager/index.js [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+    }));
+
+    SlashCommandParser.addCommandObject(SlashCommand.fromProps({
+        name: 'profile-update',
+        helpString: 'Update the selected connection profile.',
+        callback: async () => {
+            const selectedProfile = extension_settings.connectionManager.selectedProfile;
+            const profile = extension_settings.connectionManager.profiles.find(p => p.id === selectedProfile);
+            if (!profile) {
+                return '';
```
Comment (Wolfsblvt) [inline public/scripts/extensions/connection-manager/index.js:557]: No warning if a save fails? That's kind of data loss. I'd expect at *least* a console log here, if not a toast.

File: public/scripts/extensions/connection-manager/index.js [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+
+        const profileId = selectedProfile.value;
+        extension_settings.connectionManager.selectedProfile = profileId;
+        saveSettingsDebounced();
+        await renderDetailsContent(details, detailsContent);
+
+        const profile = extension_settings.connectionManager.profiles.find(p => p.id === profileId);
+
+        if (!profile) {
+            console.log(`Profile not found: ${profileId}`);
```
Comment (Wolfsblvt) [inline public/scripts/extensions/connection-manager/index.js:388]: You'll run into this log if you select the "&lt;None&gt;" option in the dropdown. Doesn't hurt, but it's a bit wrong. Should likely add specific logif for that NONE profile here.

File: public/scripts/extensions/connection-manager/index.js [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+    'instruct-state': 'Instruct Mode',
+    'instruct': 'Instruct Template',
+    'context': 'Context Template',
+    'tokenizer': 'Tokenizer',
+};
+
+/** @type {() => SlashCommandEnumValue[]} */
+const profilesProvider = () => [
+    new SlashCommandEnumValue(NONE, NONE),
+    ...extension_settings.connectionManager.profiles.map(p => new SlashCommandEnumValue(p.name, p.name, enumTypes.name, enumIcons.server)),
```
Comment (Wolfsblvt) [inline public/scripts/extensions/connection-manager/index.js]: No need to make this multiline if the name and display name are the same value. You can just set the second argument to `null`, looks cleaner.

**Review by Wolfsblvt — CHANGES_REQUESTED** [review]
#### 1. Consistent Font Awesome Icons
image

I know I am being nitpicky here. But I'd like to see consistent icon style.
"Reload" can be debated if it's the same as "Restore", but it functions very similar, so I'd say yes.
"Delete" can simply be the same trash can.
"Save as.." mas sense with the icon we use in other places, it's not really "New", as there is no actual new content created.

#### 2. Enforce Unique Names
The unique name gen thingy is good. (Though we now have the third functionality that does unique name building in a third, different way)
But you should enforce names by checking on existing profiles on save too. For both slash command and the button itself. Likely inside the save function?
It's easily overridable by just typing the same name manually, which will screw with most of your logic, going by unique names.

image

#### 3. Console Exception for Main API on Load/Refresh
Happens regularly. Even when switching between profiles.
The rest of the profile is still loaded correctly. I think you are also not passing errors correctly along to the profile selector itself. Dunno if you **want** to continue executing on errors.
Correction, also happens when switching to a CC profile.

image
image

#### 4. Long Execution, Indicator Needed
Especially with TC it seems like the commands might take a while to execute all of them, command by command.
There is no visual indication if they still run. Especially because you switch main api first, which is on the same panel, it looks like it might already be done, while the commands are still running.
I accidently switched to the AF panel and changed the context preset during testing **before** it was changed with this command.
Should add at least a visual indicator. I would even say block the UI. Similar to the QR execution in the QR editor, make a blocking transparent popup or something, that prevents **all** inputs until it is done.

Same btw for running the slash commands. They don't have a "finished" toast either.
You should likely just add a success toast at the end of switching/loading anyway. In addition to the blocking behavior explained above. The "Refresh" functionality already has a toast at the end.

#### 5. Context Templates Still Toast
image

I checked, there is no quiet flag in there yet. That's likely the reason. Also, that toast is ages old. "Auto-select" is the wrong wording anyway, as the same function is used everywhere to just *select* the template, not just for auto-select.

#### 6. What Happens On Manual Change
Maybe a bit of an opinionated point, not really a bug/issue. But if you already have the "<None>" option, wouldn't it make more sense that switching any of the things inside a profile manually (api, templates, tokenizer, etc) will set the profile to NONE, as you currently don't really have a profile selected anymore? You manually modified your stuff.
I can see that this would be an issue with changing and overwriting a profile though.
Maybe leaving it as is is good enough.
People were already asking for "dirty" indicators on the templates/sampler settings too. Maybe this is something to be done for all of those at once.

#### Additional Stuff / Questions
1. If someone deletes a sampler preset or instruct/context, the profile will not notice this. Should be fine - I mean kinda your fault. At least you still get a toast.
2. Would be nice if the `/profile` slash command hat an `await` argument too, to be blocking until the profile is loaded.

File: public/scripts/slash-commands.js (hunk 1 above)
Comment (Cohee1207) [reply public/scripts/slash-commands.js]: It wasn't working for whatever reason in staging. Pretty strange that now it does. But it's good to have it here explicitly set to true.

**Review by Wolfsblvt — APPROVED** [review]
Two last small nitpicks that I fixed.
Is only marginally related to this, but I mentioned it here in the branch, and you changed/used the quiet flags anyway, so I added this here.
Second one is disabling the buttons that you can't use, because... duh.

Tested the PR one last time and couldn't find anything else.
Looks very nice.

People will be so so happy.
(I'll tell anyone those are not preset-presets, I promise.)

**Pull request merged**